proposals/stdlib/uuid-additional-versions.md

```diff
@@ -0,0 +1,263 @@
+ensures the API is consistent with other Kotlin stdlib API, and ensures it has a good performance.
+
+## Proposal
+
+It is proposed to support two additional version of UUID as specified in RFC 9562(https://www.rfc-editor.org/rfc/rfc9562.txt).
+
+The earlier proposed v6 version is not included, as it was never widely adopted and fully superseded by v7.
+
+### Proposed API
+
+#### UUID generator interface
+
+A new interface, representing a UUID generator is added. Although a pure function-based API is often simpler to use,
+an object represents the stateful nature of UUID generation better. It is also more handy in the following cases:
+
+* Providing a repeatable way to generate UUIDs, for example when used in testing.
+* Tune the UUID generation process, for example changing the entropy of generated UUIDs.
+* Provide a pooling implementation (see UUID pooling).
+
+```kotlin
+    fun interface UuidGenerator {
+    
+        fun generate() : Uuid
+        
+   }
+```
+
+A separate interface is created for name-based UUID generators, as they need to accept input to generate UUIDs.
+
+```kotlin
+
+
+/**
+ * Namespace used for DNS names.
+ */
+val NAMESPACE_DNS: Uuid = Uuid.parse("6ba7b810-9dad-11d1-80b4-00c04fd430c8")
+
+/**
+ * Namespace used for URLs.
+ */
+val NAMESPACE_URL: Uuid = Uuid.parse("6ba7b811-9dad-11d1-80b4-00c04fd430c8")
+/**
+ * Namespace used for names created from OID.
+ */
+val NAMESPACE_OID: Uuid = Uuid.parse("6ba7b812-9dad-11d1-80b4-00c04fd430c8")
+
+/**
+ * Namespace used for X.500 identifier names
+ */
+val NAMESPACE_X500: Uuid = Uuid.parse("6ba7b814-9dad-11d1-80b4-00c04fd430c8")
+
+public interface NameBasedUuidGenerator {
+
+    public fun generate(name: CharSequence): Uuid
+
+    public fun generate(name: ByteArray): Uuid
+
+}
+```
+
+
+#### Companion object of class `Uuid` new methods
+
+New methods are added to the Companion object of class `Uuid`. These methods will provide the implementations
+of the new generators as well as an implementation for already supported UUID v4.
+
+```kotlin
+
+/**
+ *  Returns a UUIDv4 generator, that generates random UUIDs.
+ */
+public fun uuidV4Generator(): UuidGenerator
+
+/**
+ *  Returns a UUIDv5 generator, that generates UUIDs from supplied names
+ */
+public fun uuidV5Generator(namespace: Uuid = Uuid.NIL): NameBasedUuidGenerator
+
+/**
+ *  Returns a UUIDv7 generator, that generates time-ordered UUIDs.
+ */
+public fun uuidV7Generator(): UuidGenerator
+
+```
+
+#### Code examples
+
+```kotlin
+
+import kotlin.uuid.Uuid
+import kotlin.uuid.NAMESPACE_URL
+
+fun main() {
+
+    val uuidV5Generator = Uuid.uuidV5Generator(NAMESPACE_URL)
+    val uuidV5 = oneMoreV5Generator.generate("http://www.jetbrains.com")
+
+    val uuidV7Generator = Uuid.uuidV7Generator()
+    val uuidV7_1 = uuidV7Generator.generate()
+    val uuidV7_2 = uuidV7Generator.generate()
+}
+
+```
+
+## Dependencies
+
+There should be no new dependencies for JVM platform.
+TODO for other runtimes.
```

Comment:
Would it be possible to generate v5 UUIDs on platforms other than JVM?
For example, on Wasm/WASI?

Reply:
@fzhinkin the main challenge would be finding a suitable SHA-1 implementation.
A quick search shows that there are widely used implementations with compatible licenses.

For example, https://github.com/RustCrypto/hashes can be used for WebAssembly.

P.S. This opens the very interesting topic of having crypto primitives in Kotlin stblib. IMHO, there are very much needed

Reply:
> At the moment, Libraries team won't be able to drive that process. So, unfortunately, chances of including the SHA1 implementation are diminishingly small.

I would not suggest a home-grown implementation. Rather, I would use a well-established library similar to how Firefox uses NSS, OpenSSH uses LibreSSL and plethore of other projects use BouncyCastle